File: moq-web/src/internal/message/track_test.ts
import { assertEquals } from "@std/assert";
import { TrackMessage } from "./track.ts";
import { Buffer } from "@okdaichi/golikejs/bytes";

Deno.test("TrackMessage - encode/decode roundtrip", async (t) => {
	const testCases = {
		"valid message": { broadcastPath: "/live/alice", trackName: "video" },
		"empty fields": { broadcastPath: "", trackName: "" },
	};

	for (const [caseName, input] of Object.entries(testCases)) {
		await t.step(caseName, async () => {
			const buffer = Buffer.make(128);
			const message = new TrackMessage(input);
			assertEquals(await message.encode(buffer), undefined);

			const decoded = new TrackMessage({});
			assertEquals(await decoded.decode(buffer), undefined);
			assertEquals(decoded.broadcastPath, input.broadcastPath);
			assertEquals(decoded.trackName, input.trackName);
		});
	}

	await t.step("decode returns error on empty input", async () => {
		const buffer = Buffer.make(0);
		const decoded = new TrackMessage({});
		const err = await decoded.decode(buffer);
		assertEquals(err !== undefined, true);
	});
});

File: moqt/internal/message/track_test.go
package message_test

import (
	"bytes"
	"testing"

	"github.com/qumo-dev/gomoqt/moqt/internal/message"
	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
)

func TestTrackMessage_EncodeDecode(t *testing.T) {
	tests := map[string]struct {
		input message.TrackMessage
	}{
		"valid message": {
			input: message.TrackMessage{
				BroadcastPath: "/live/alice",
				TrackName:     "video",
			},
		},
		"empty fields": {
			input: message.TrackMessage{},
		},
	}

	for name, tt := range tests {
		t.Run(name, func(t *testing.T) {
			var buf bytes.Buffer

			require.NoError(t, tt.input.Encode(&buf))

			var decoded message.TrackMessage
			require.NoError(t, decoded.Decode(&buf))

			assert.Equal(t, tt.input, decoded)
		})
	}
}

func TestTrackMessage_DecodeErrors(t *testing.T) {
	var decoded message.TrackMessage
	err := decoded.Decode(bytes.NewReader([]byte{}))
	assert.Error(t, err)
}
